### PR TITLE
Redshift - Copy to temp table, prune table based on date, utilization of init_copy and post_copy

### PR DESCRIPTION
**Pull request by dlstadther**
Added support for redshift copy to temp table, prune table (delete some amount of records based on date), init copy, post copy (also added to rdbms.py)

Also includes the rdbms.py post_copy update found in PR #1313

### CODE REVIEW

**Comment by Tarrasch** [comment]
Can you either keep the PRs seperate or close #1313?

File: luigi/contrib/redshift.py [hunk 1]
```diff
@@ -103,6 +103,15 @@ def copy_options(self):
         """
         return ''
 
+    def prune_table(self):
```
Comment (Tarrasch) [inline luigi/contrib/redshift.py]: Docstrings would be nice :)

**Comment by Tarrasch** [comment]
It would be great if you can find somebody who uses redshift in luigi to review. Thanks :)

**Comment by dlstadther** [comment]
I closed PR #1313 and will include updates here.

Luigi redshift users are quiet and unknown. Mortar Data uses some redshift functionality in their implementation of luigi, but they seem to have been inactive in the luigi community for awhile.

**Comment by dlstadther** [comment]
As rdbms.py is a common module for which is to be implemented (not executed itself), I do not see how it needs its own tests. Postgres.py and Redshift.py both implement Rdbms.py; thus, their unit tests indirectly test rdbms.py.

File: test/contrib/redshift_test.py [hunk 2]
```diff
@@ -44,10 +43,42 @@ class DummyS3CopyToTable(luigi.contrib.redshift.S3CopyToTable):
...
+        ('some_text', 'text'),
+        ('some_int', 'int'),
+    )
+
+    aws_access_key_id = AWS_ACCESS_KEY
+    aws_secret_access_key = AWS_SECRET_KEY
+    copy_options = ''
+    sql = ["insert into dummy_table select * from stage_dummy_table;"]
+    prune_date = 'current_date - 30'
+    prune_column = 'dumb_date'
```
Comment (Tarrasch) [inline test/contrib/redshift_test.py]: This looks weird to me. Is `prune_column` (as well as the other 2) a method or a property? I mean, is it `myobj.prune_column` or `myobj.prune_column()` that's supposed to be a string?

**Comment by Tarrasch** [comment]
@dlstadther, oh ok. I didn't know that redshift inherited rdbms. If you think all code paths are tested. I'm fine with that :)

File: test/contrib/redshift_test.py (hunk 2 above)
Comment (dlstadther) [inline test/contrib/redshift_test.py]: I intend them to be properties...optional properties. If I specify @property over them in redshift.py, will that allow them to be optional properties with 'None' as their default value?

Comment (dlstadther) [inline test/contrib/redshift_test.py]: @Tarrasch Do you have any suggestions on how to specify prune_* as optional properties such that they can be used like the test case?

Comment (Tarrasch) [inline test/contrib/redshift_test.py]: Can't you just do:

```
@property
def prune_table(self):
    return None
```

Or maybe even just `prune_table = None`? Hmm...

Comment (dlstadther) [inline test/contrib/redshift_test.py]: @Tarrasch If the `prune_*` variables should be marked as properties, shouldn't `table_type` and `table_attributes` also be marked as such?

**Comment by moandcompany** [comment]
I'm currently using Luigi with Redshift, so I'll try these out when I get a chance.

File: luigi/contrib/rdbms.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 # limitations under the License.
 #
 """
-A common module for postgres like databases, such as postgres or redshift
+A common module for posgres like databases, such as postgres or redshift
```
Comment (Tarrasch) [inline luigi/contrib/rdbms.py]: huh? Introducing a spell error?

Comment (dlstadther) [inline luigi/contrib/rdbms.py]: I'm not sure how that happened. I will fix and merge commits.

File: test/contrib/redshift_test.py (hunk 2 above)
Comment (Tarrasch) [inline test/contrib/redshift_test.py]: Hmm. Good point, maybe it's better to stay consistent. Do something clever. :)

**Comment by dlstadther** [comment]
+bump

(PRs are getting a little stagnate...I understand Luigi is undergoing a change in management, but the development must go on!)

Comments, questions, concerns, etc. are welcome.

**Comment by dlstadther** [comment]
@erikbern Thanks for the quick action! Greatly appreciated!

**Comment by erikbern** [comment]
yes no problem – I'm happy to merge most stuff that goes into contrib right away :) just forgot about it

let me check if other things are pending

**Pull request merged**